addon/resolver.js

```diff
@@ -284,23 +285,35 @@ var Resolver = DefaultResolver.extend({
       return moduleName;
     } else if (this._moduleRegistry.has(underscoredModuleName)) {
       return underscoredModuleName;
-    } else {
-      // workaround for dasherized partials:
-      // something/something/-something => something/something/_something
-      var partializedModuleName = moduleName.replace(/\/-([^\/]*)$/, '/_$1');
-
-      if (this._moduleRegistry.has(partializedModuleName)) {
-        Ember.deprecate('Modules should not contain underscores. ' +
-                        'Attempted to lookup "'+moduleName+'" which ' +
-                        'was not found. Please rename "'+partializedModuleName+'" '+
-                        'to "'+moduleName+'" instead.', false,
-                        { id: 'ember-resolver.underscored-modules', until: '3.0.0' });
-
-        return partializedModuleName;
-      } else {
-        return moduleName;
-      }
     }
+
+    // workaround for dasherized partials:
+    // something/something/-something => something/something/_something
+    var partializedModuleName = moduleName.replace(/\/-([^\/]*)$/, '/_$1');
+
+    if (this._moduleRegistry.has(partializedModuleName)) {
+      Ember.deprecate('Modules should not contain underscores. ' +
+                      'Attempted to lookup "'+moduleName+'" which ' +
+                      'was not found. Please rename "'+partializedModuleName+'" '+
+                      'to "'+moduleName+'" instead.', false,
+                      { id: 'ember-resolver.underscored-modules', until: '3.0.0' });
+
+      return partializedModuleName;
+    }
+
+    var isCamelCase = moduleName.match(/[a-z]+[A-Z]+/);
+    var dasherizedModuleName = dasherize(moduleName);
+    if (isCamelCase && this._moduleRegistry.has(dasherizedModuleName) && (this._camelCaseModuleWarning.indexOf(parsedName.fullName) === -1)) {
```

Comment:
We should restructure to completely avoid any work in production mode (I believe that would require moving this into a runInDebug)